### PR TITLE
add chp 1

### PR DESCRIPTION
**Pull request by KevalJoshi**

### CODE REVIEW

File: A&S/OneAway.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+function IfTheyAreOneEditAway(str1,str2){
```
Comment (milincjoshi) [inline A&S/OneAway.js:1]: what if both the strings passed are empty?

File: A&S/OneAway.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                var strln1 = str1.length;
+                var strln2 = str2.length;
+                if((strln1 - strln2) > 1){
+                    return false;
+                }
+                let i=0 , j=0, count = 0;
```
Comment (milincjoshi) [inline A&S/OneAway.js:7]: What does i and j represent?

File: A&S/OneAway.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                while(i<strln1 && j<strln2){
+                    if(str1[i] != str2[j]){
+                        if(count == 1){
```
Comment (milincjoshi) [inline A&S/OneAway.js:10]: what does count represent?

File: A&S/OneAway.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                            i++;
+                            j++;
+                        }
+                }
+                        if (i < strln1 || j < strln2){
+                        count++;
+                        }
+                return count == 1;
+            }
+           console.log(IfTheyAreOneEditAway("pale","ple"));
```
Comment (milincjoshi) [inline A&S/OneAway.js:35]: can we add test cases for both empty strings, and one of the strings empty?

File: A&S/OneAway.js (hunk 1 above)
Comment (milincjoshi) [inline A&S/OneAway.js:1]: can we use proper format and add comments?

File: A&S/PalindromePermutation.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+  function isPalindrome(str){
```
Comment (milincjoshi) [inline A&S/PalindromePermutation.js:1]: can we add a test case with even number of characters and being a palindrome?

Comment (milincjoshi) [inline A&S/PalindromePermutation.js:1]: Same comment for adding comments and formatting for all files.

File: A&S/RotateMatrix.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+ var N = 4;
```
Comment (milincjoshi) [inline A&S/RotateMatrix.js:1]: can we also add test cases where the matrix shape is not a square? i.e N!=4?

File: A&S/RotateMatrix.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+            function rotate90Clockwise(arr){
+                for (j = 0; j < N; j++) {
```
Comment (milincjoshi) [inline A&S/RotateMatrix.js:3]: what if the matrix is empty?

File: A&S/StringCompression.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+ function StringCompression(str){
```
Comment (milincjoshi) [inline A&S/StringCompression.js:1]: can we also implement a string compressor in-place?

File: A&S/URLify.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        strArr[newLength - 1] = strArr[i];
+           newLength = newLength - 1;
+       }
+   }
+   
+   return strArr;
+}
+   
+
+        var replace = replaceString("Mr John Smith is liar        ",21);
```
Comment (milincjoshi) [inline A&S/URLify.js:31]: Add a few more test cases to test this.

File: A&S/ZeroMatrix.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+                    }else{
+                        for(j = 0; j < matrix.length; j++){
+                            matrix[j][index[i]] = 0;
+                        }
+                    }
+                }
+                return matrix;
+            }
+            var matrix = [[ 1, 2, 3, 4 ],
+                          [ 5, 6, 7, 8 ],
```
Comment (milincjoshi) [inline A&S/ZeroMatrix.js:26]: Add test case for 0 as well.

File: A&S/checkPermutation.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+function checkPermutation(str1, str2){
+
+    if(str1.length != str2.length){
```
Comment (milincjoshi) [inline A&S/checkPermutation.js:3]: implement same function without sort as well.

File: A&S/isUnique.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    }
+    else{
+        return false;
+    }
+}
+  
+  const hasUniqueCharacters = isUnique("milin");
+  console.log(hasUniqueCharacters);
+
+  // Time Complexity of the above code is O(n) where is the length of the input string
```
Comment (milincjoshi) [inline A&S/isUnique.js:16]: Add Space and Time complexity for all of above.

File: A&S/isUnique.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+
+function isUnique(str) {
+    var strSet = new Set(str);
```
Comment (milincjoshi) [inline A&S/isUnique.js:3]: what if this line is not available as a utility to be used? How else can we implement the same function?
